### PR TITLE
Support SSH for go get

### LINKED ISSUE #12192: Support SSH for go get
- Gitea version (or commit ref): 1.12.1 built with GNU Make 4.2.1, go1.14.4 : bindata, sqlite, sqlite_unlock_notify  
- Git version: 2.24.3
- Operating system: Arch Linux
- Database (use `[x]`):
  - [ ] PostgreSQL
  - [ ] MySQL
  - [ ] MSSQL
  - [x] SQLite
- Can you reproduce the bug at https://try.gitea.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log gist:

## Description
We have a private gitea repository containing a go module. When we want to `go get git.company.com/organization/private`, gitea returns only the HTTPS link of the repository, not the SSH one.  
To reproduce, one can simply open a link in the browser: `https://git.company.com/organization/private?go-get=1`  
and examine the returned HTML:  
```html
<!DOCTYPE html>
<html>
<head>
<meta name="go-import" content="git.company.com/organization/private git https://git.company.com/organization/private.git">
<meta name="go-source" content="git.company.com/organization/private _ https://git.company.com/organization/private/src/branch/master{/dir} https://git.company.com/organization/private/src/branch/master{/dir}/{file}#L{line}">
</head>
<body cz-shortcut-listen="true">
go get git.company.com/organization/private
</body>
</html>
```

Here are the relevant source code locations:
```
templates/base/head.tmpl:	<meta name="go-import" content="{{.GoGetImport}} git {{.CloneLink.HTTPS}}">
modules/context/repo.go:	ctx.PlainText(200, []byte(com.Expand(`<meta name="go-import" content="{GoGetImport} git {CloneLink}">`,
modules/context/context.go:		<meta name="go-import" content="{GoGetImport} git {CloneLink}">
```
One can see in these files/lines that only the HTTPS path is ever returned. We expected that this can be at least configured to either return the SSH or HTTPS one. When disabling HTTPS for the repo, the returned links still stayed the same.

### PR DESCRIPTION
**Pull request by mei-rune**
fix #12192 Support SSH for go get

### CODE REVIEW

**Comment by wxiaoguang** [comment]
Awesome, thank you very much!

The only nits seems:

1. The config name "USE_GIT_CLONE_URL" seems not related to its meaning?  It might be something like "GO_GET_CLONE_URL_PROTOCOL = ssh/https" (only check "ssh" for ssh, other values for http/https?)
2. The `app.example.ini` and `cheatsheet-....-en.md` needs to be updated accordingly.


Thank you again ~~

File: modules/context/repo.go [hunk 1]
```diff
@@ -319,7 +319,14 @@ func EarlyResponseForGoGetMeta(ctx *Context) {
 		ctx.PlainText(http.StatusBadRequest, "invalid repository path")
 		return
 	}
-	goImportContent := fmt.Sprintf("%s git %s", ComposeGoGetImport(username, reponame), repo_model.ComposeHTTPSCloneURL(username, reponame))
+
+	var cloneURL string
+	if setting.Repository.UseGitCloneURL {
```
Comment (lunny) [inline modules/context/repo.go]: Are there any go get parameter to indicate a ssh or a http protocol?

Comment (mei-rune) [reply modules/context/repo.go]: yes

Comment (wxiaoguang) [reply modules/context/repo.go]: I haven't looked into details nor used "go get over ssh", so I'm not sure whether my understanding is correct, according to:

1. the previous comment https://github.com/go-gitea/gitea/issues/12192#issuecomment-657190979
2. the fact that how `go get` works: it only reads the `github.com/owner/repo` in `go.mod`: there is no chance for `go get` to pass some "ssh/https" parameter.

So, I would assume this PR's approach is right. Correct me if I am wrong.

**Review by wxiaoguang — APPROVED** [review]

**Comment by jolheiser** [comment]
I won't necessarily hold back this PR, but the more common way to do this is to handle it in your git config. 
```
[url "ssh://git@domain.tld/"]
    insteadOf = https://domain.tld/
```
While it's perhaps nice for Gitea to be able to do this, it could cause confusion when this doesn't work on other forges that don't have a knob for it.

**Comment by jolheiser** [comment]
`go get` just uses git under the hood (when git is the vcs specified), so it uses this meta tag to pass to `clone`.
From there you're (more or less) in git land.

**Comment by wxiaoguang** [comment]
The `url insteadOf` trick is pretty good, but there is one limitation: if a site admin deploys Gitea in their internal network, they are able to force every developer to do so ..... 🤣

**Comment by silverwind** [comment]
I see, so the current flow is:

- go get to HTTPS -> gets HTTPS url, git clone changes HTTPS to SSH

Afterwards with the option flipped:

- go get to HTTPS -> gets SSH url, git clone SSH, skipping the requirement for config

File: modules/context/repo.go (hunk 1 above)
Comment (lunny) [reply modules/context/repo.go]: ~Even if we need a new configuration, the name `UseGitCloneURL` is still confusing. Maybe `GoGetProtocol` is a better name?~

**Comment by lunny** [comment]
Take a look at the go get protocol, Go client will try different protocol one by one. So maybe we can put both. That means, both meta will be included so which one is work will be decided by Go client.

**Comment by wxiaoguang** [comment]
Yup, if someone could spend some time on confirming that "putting both SSH/HTTPS links" works, that's much better, and no need to add the new config option.

**Comment by lunny** [comment]
At least, go get will read multiple metas if I'm not wrong. https://github.com/golang/go/blob/a25688d406f2a4296d39b22a2b10aea7178eddc6/src/cmd/go/internal/vcs/discovery.go#L31

**Comment by wxiaoguang** [comment]
> At least, go get will read multiple metas if I'm not wrong. https://github.com/golang/go/blob/a25688d406f2a4296d39b22a2b10aea7178eddc6/src/cmd/go/internal/vcs/discovery.go#L31

Does it go back to the history? "the previous comment https://github.com/go-gitea/gitea/issues/12192#issuecomment-657190979"

image

File: modules/setting/repository.go [hunk 2]
```diff
@@ -273,6 +274,7 @@ func loadRepositoryFrom(rootCfg ConfigProvider) {
 	sec := rootCfg.Section("repository")
 	Repository.DisableHTTPGit = sec.Key("DISABLE_HTTP_GIT").MustBool()
 	Repository.UseCompatSSHURI = sec.Key("USE_COMPAT_SSH_URI").MustBool()
+	Repository.GoGetCloneURLProtocol = sec.Key("GO_GET_CLONE_URL_PROTOCOL").MustString("")
```
Comment (lunny) [inline modules/setting/repository.go]:
```suggestion
	Repository.GoGetCloneURLProtocol = sec.Key("GO_GET_CLONE_URL_PROTOCOL").MustString("https")
```

File: routers/web/goget.go [hunk 3]
```diff
@@ -66,7 +66,14 @@ func goGet(ctx *context.Context) {
 	}
 
 	goGetImport := context.ComposeGoGetImport(ownerName, trimmedRepoName)
-	goImportContent := fmt.Sprintf("%s git %s", goGetImport, repo_model.ComposeHTTPSCloneURL(ownerName, repoName) /*CloneLink*/)
+	
+	var cloneURL string
+	if setting.Repository.GoGetCloneURLProtocol == "ssh" {
```
Comment (lunny) [inline routers/web/goget.go:71]: ~You missed modules/context/repo.go:322~

Comment (wxiaoguang) [reply routers/web/goget.go:71]: What do you mean? can't understand.

**Review by lunny — APPROVED** [review]

File: templates/repo/issue/list.tmpl [hunk 4]
```diff
@@ -38,6 +38,7 @@
 								<input type="text" placeholder="{{.locale.Tr "repo.issues.filter_label"}}">
 							</div>
 							<span class="info">{{.locale.Tr "repo.issues.filter_label_exclude" | Safe}}</span>
+							<a class="{{if $.MilestoneID}}{{if eq $.MilestoneID -1}}active selected {{end}}{{end}}item" href="{{$.Link}}?q={{$.Keyword}}&type={{$.ViewType}}&sort={{$.SortType}}&state={{$.State}}&labels={{.SelectLabels}}&milestone=-1&project={{$.ProjectID}}&assignee={{$.AssigneeID}}&poster={{$.PosterID}}">{{.locale.Tr "repo.issues.filter_milestone_none"}}</a>
```
Comment (wxiaoguang) [inline templates/repo/issue/list.tmpl]: What's this? @lunny

Comment (lunny) [reply templates/repo/issue/list.tmpl]: :(

**Comment by mei-rune** [comment]
only "db-tests / test-mysql8 (pull_request) " fail, but This has nothing to do with this patch. 

只剩一个 "db-tests / test-mysql8 (pull_request) ", 但它与本补丁无关

**Review by wxiaoguang — APPROVED** [review]

**Review by wolfogre — APPROVED** [review]

**Pull request merged**